File: framework/kotlin/src/main/kotlin/soda/kotlin/web/TimeLimitedWork.kt
package soda.kotlin.web

import java.util.concurrent.FutureTask

class TimeLimitedWork(private val job: () -> String) {

    private var jobThread: Thread? = null

    fun start(): FutureTask<String> {
        return FutureTask { job() }.also {
            jobThread = Thread(it)
            jobThread!!.start()
        }
    }

    fun kill() {
        try {
            jobThread?.stop()
        } catch (th: Throwable) {
            th.printStackTrace()
        }
    }
}

File: framework/scala/src/main/scala/soda/scala/web/TimeLimitedWork.scala
package soda.scala.web

import java.util.concurrent.FutureTask

class TimeLimitedWork(job: () => String) {

  private var jobThread: Thread = null

  def start(): FutureTask[String] = {
    val task = new FutureTask[String](() => job())
    jobThread = new Thread(task)
    jobThread.start()
    task
  }

  def kill(): Unit = {
    try {
      if (jobThread != null) {
        jobThread.stop()
      }
    } catch {
      case th: Throwable => th.printStackTrace()
    }
  }
}
